### PR TITLE
Mac app extra binary signing tweaks

### PR DESCRIPTION
**Pull request by palerdot**
### Problem

Mac app signing fails when extra binaries are included in the app bundle. [`Simple Backup`](https://github.com/JackGruber/joplin-plugin-backup) plugin uses `7zip` binary for its functionality. When bundling this plugin as a default plugin, mac app signing fails because the `7zip` binary is not signed by current build process. Related to https://github.com/laurent22/joplin/issues/7934

### Solution

This PR tries to solve the issue by doing the following.
- Updrading to latest `@electron/osx-sign@v1.x` version from `electron-osx-sign@v0.6.0`. 
- Explicitly specifying the binaries to be signed  for `@electron/osx-sign` - https://www.electron.build/configuration/mac.html, https://github.com/electron/osx-sign#usage.

Current joplin app has `electron-osx-sign@v0.6.0` as a transitive/indirect dependency included by `electron-builder@v23`. As per this github issue comment - https://github.com/electron-userland/electron-builder/issues/4656#issuecomment-611639310, `electron-osx-sign` made a patch previously addressing the binary signing issue. This PR upgrades to `electron-builder@v24`, which bundles the latest package/namespace `@electron/osx-sign@v1.x` as its dependency. 

After upgrading the package, the listed `osx-sign` package dependency listed by `yarn why` is 
```shell
➜ yarn why @electron/osx-sign
└─ app-builder-lib@npm:24.1.2
   └─ @electron/osx-sign@npm:1.0.4 (via npm:^1.0.4)
```

### Testing

After upgrading the package, package was built on a mac machine manually and it succeded. It is not possible to locally sign the app package because of lack of a valid developer credentials. For more details, please check the log below
```shell
  • skipped macOS application code signing  reason=cannot find valid "Developer ID Application" identity or custom non-Apple code signing certificate, it could cause some undefined behaviour, e.g. macOS localized description not visible, see https://electron.build/code-signing allIdentities=     0 identities found
                                                Valid identities only
     0 valid identities found
```

This PR has to be tested with existing CI setup or a local machine, which has valid mac developer credentials.

### Miscellaneous

- `electron-builder@v24.1.x` is currently published with `next` tag on npm - https://www.npmjs.com/package/electron-builder?activeTab=versions. The download count suggests that people are actively using `24.x`. In case this PR works, but only `latest` tag version has to be used, this PR has to wait till `v24` is bumped to `latest` tag.
- `@electron/notarize` is the new recommended namespace and this PR upgrades to this new namespace from `electron-notarize`.
- If this PR succeeds, then future additional binaries may have to be included in the `mac.binaries` field in the `package.json`.
- `@electron/rebuild` is the new recommended namespace and this PR upgrades to this new namespace from `electron-rebuild`. One of the warnings when building with `electron-builder` is `@electron/rebuild is already incorporated into electron-builder, please consider to remove excess dependency from devDependencies`. Maybe, this dependency has to be removed in the future. For now, the package is upgraded to latest version matching the internal dependency of electron-builder.

### CODE REVIEW

**Comment by laurent22** [comment]
Thanks Arun, but it is failing when signing. Maybe an incorrect path?

      ⨯ Command failed: codesign --sign Developer ID Application: Laurent Cozic (***) --force --keychain /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/7a64b6fec4be4f0728c020de941132d7d46be21adeb95fb96be9c2ea5a71e786.keychain --timestamp --options runtime --entitlements /Users/runner/work/joplin/joplin/packages/app-desktop/node_modules/app-builder-lib/templates/entitlements.mac.plist /Users/runner/work/joplin/joplin/packages/app-desktop/dist/mac/Joplin.app/build/defaultPlugins/io.github.jackgruber.backup/plugin.jpl/7zip-bin/mac/x64/7za
    /Users/runner/work/joplin/joplin/packages/app-desktop/dist/mac/Joplin.app/build/defaultPlugins/io.github.jackgruber.backup/plugin.jpl/7zip-bin/mac/x64/7za: No such file or directory
      failedTask=build stackTrace=Error: Command failed: codesign --sign Developer ID Application: Laurent Cozic (***) --force --keychain /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/7a64b6fec4be4f0728c020de941132d7d46be21adeb95fb96be9c2ea5a71e786.keychain --timestamp --options runtime --entitlements /Users/runner/work/joplin/joplin/packages/app-desktop/node_modules/app-builder-lib/templates/entitlements.mac.plist /Users/runner/work/joplin/joplin/packages/app-desktop/dist/mac/Joplin.app/build/defaultPlugins/io.github.jackgruber.backup/plugin.jpl/7zip-bin/mac/x64/7za
    /Users/runner/work/joplin/joplin/packages/app-desktop/dist/mac/Joplin.app/build/defaultPlugins/io.github.jackgruber.backup/plugin.jpl/7zip-bin/mac/x64/7za: No such file or directory
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    at ChildProcess.exithandler (node:child_process:419:12)
        at ChildProcess.emit (node:events:513:28)
        at maybeClose (node:internal/child_process:1091:16)
        at Socket.<anonymous> (node:internal/child_process:449:11)
        at Socket.emit (node:events:513:28)
        at Pipe.<anonymous> (node:net:322:12)
    From previous event:
        at processImmediate (node:internal/timers:476:21)
    From previous event:
        at readDirectoryAndSign (/Users/runner/work/joplin/joplin/packages/app-desktop/node_modules/app-builder-lib/src/macPackager.ts:466:29)
        at MacPackager.signApp (/Users/runner/work/joplin/joplin/packages/app-desktop/node_modules/app-builder-lib/src/macPackager.ts:476:11)
        at MacPackager.doSignAfterPack (/Users/runner/work/joplin/joplin/packages/app-desktop/node_modules/app-builder-lib/src/platformPackager.ts:332:21)
        at MacPackager.doPack (/Users/runner/work/joplin/joplin/packages/app-desktop/node_modules/app-builder-lib/src/platformPackager.ts:317:7)
        at MacPackager.pack (/Users/runner/work/joplin/joplin/packages/app-desktop/node_modules/app-builder-lib/src/macPackager.ts:194:7)
        at Packager.doBuild (/Users/runner/work/joplin/joplin/packages/app-desktop/node_modules/app-builder-lib/src/packager.ts:442:9)
        at executeFinally (/Users/runner/work/joplin/joplin/packages/app-desktop/node_modules/builder-util/src/promise.ts:12:14)
        at Packager._build (/Users/runner/work/joplin/joplin/packages/app-desktop/node_modules/app-builder-lib/src/packager.ts:376:31)
        at Packager.build (/Users/runner/work/joplin/joplin/packages/app-desktop/node_modules/app-builder-lib/src/packager.ts:337:12)
        at executeFinally (/Users/runner/work/joplin/joplin/packages/app-desktop/node_modules/builder-util/src/promise.ts:12:14)
    Error: Process completed with exit code 1.

**Comment by laurent22** [comment]
Seems to work now, please update the run_ci script

File: packages/app-desktop/gulpfile.js [hunk 1]
```diff
@@ -35,6 +36,13 @@ const tasks = {
 			);
 		},
 	},
+	bundleDefaultPlugins: {
+		fn: async () => {
+			const toolsDirectory = `${__dirname}/../tools`;
+			process.chdir(toolsDirectory);
```
Comment (laurent22) [inline packages/app-desktop/gulpfile.js]: that shouldn't be here. The script itself should know from which directory it needs to run, so any change of directory should be in the script

File: packages/tools/bundleDefaultPlugins.ts [hunk 2]
```diff
@@ -94,3 +94,12 @@ if (require.main === module) {
 		process.exit(1);
 	});
 }
+
+module.exports = async function bundleDefaultPlugins() {
```
Comment (laurent22) [inline packages/tools/bundleDefaultPlugins.ts]: Please read coding_style.md, we should only use arrow functions in new code

File: packages/tools/bundleDefaultPlugins.ts [hunk 3]
```diff
@@ -94,3 +94,12 @@ if (require.main === module) {
...
+	try {
+		await start();
+	}	catch (e) {
+		console.error('Error bundling default plugins ', e);
+		process.exit(1);
+	}
+};
```
Comment (laurent22) [inline packages/tools/bundleDefaultPlugins.ts]: All that indirection shouldn't be needed. If the script fails, gulp handles the error

**Comment by palerdot** [comment]
Closing in favour of https://github.com/laurent22/joplin/pull/8040

The issue is with https://github.com/laurent22/joplin/blob/dev/packages/utils/splitCommandString.ts#L3, which is causing problem in windows by removing all the `\\` when run as part of `yarn run` (`or gulp bundleDefaultPlugins`). Changes to build setup/TS gulpfile and other script changes are not needed. 

If these changes are really needed this can be reopened.

PS: Mac signing still fails, but it is a totally different problem as it worked when ran via `run_ci.sh`

**Comment by palerdot** [comment]
Also it seems the commit for which CI which passed didn't actually sign the mac app - https://github.com/laurent22/joplin/actions/runs/4628647013/jobs/8187988538.

**Pull request closed without merging**